spec/lang/step/operators.md

```diff
@@ -131,18 +134,46 @@ impl<M: Memory> Machine<M> {
 impl<M: Memory> Machine<M> {
     fn eval_un_op(&self, UnOp::ComputeSize(ty): UnOp, (operand, op_ty): (Value<M>, Type)) -> Result<(Value<M>, Type)> {
         let meta = PointerMeta::from_value::<M>(operand);
-        let size = ty.layout::<M::T>().compute_size(meta);
+        let size = ty.layout::<M::T>().compute_size(meta, self.vtable_lookup());
         ret((Value::Int(size.bytes()), Type::Int(IntType::usize_ty::<M::T>())))
     }
 
     fn eval_un_op(&self, UnOp::ComputeAlign(ty): UnOp, (operand, op_ty): (Value<M>, Type)) -> Result<(Value<M>, Type)> {
         let meta = PointerMeta::from_value::<M>(operand);
-        let align = ty.layout::<M::T>().compute_align(meta);
+        let align = ty.layout::<M::T>().compute_align(meta, self.vtable_lookup());
         ret((Value::Int(align.bytes()), Type::Int(IntType::usize_ty::<M::T>())))
     }
 }
 ```
 
+### VTable Lookups
+
+Dynamic dispatch in MiniRust is represented as a `Call` to the result of an explicit `VTableMethodLookup` expression.
+This expression works on vtable pointers, which can be extracted by `GetMetadata`.
+Which method is invoked is represented by the `method` parameter, which corresponds to a function of the trait.
+
+```rust
+impl<M: Memory> Machine<M> {
+    fn eval_un_op(&self, UnOp::VTableMethodLookup(method): UnOp, (operand, op_ty): (Value<M>, Type)) -> Result<(Value<M>, Type)> {
+        let (Value::Ptr(ptr), Type::Ptr(_ptr_ty)) = (operand, op_ty) else {
+            panic!("vtable lookup on non-pointer");
+        };
+        // It is checked in check_value that the vtable is always valid.
+        let vtable = self.vtable_lookup()(ptr.thin_pointer);
+        let Some(fn_name) = vtable.methods.get(method) else {
+            // This would be a type error, but since we don't store the trait, we do not type check this.
```

Comment:
```suggestion
            // This would be a type error, but since we don't store the trait in the VTabePtr type,
            // we do not type check this.
            // FIXME: Seems like it'd be nice to store the trait there?
```

Reply:
This is a disadvantage of having `VTableMethodLookup` operate on the metadata (i.e. the vtable pointer) directly. If we had it on the trait object pointer, the type would include the vtable name for the metadata.
So we could have `VTableMethodLookup` take the trait in addition to the method (which was once effectively the case with globally unique MethodNames).

Reply:
I think the fix is to make `PtrType::VTabePtr` carry the trait name. But that can be a future PR.